ltm/inserted_context_agent.py

```diff
@@ -0,0 +1,577 @@
+import json
+import os
+import uuid
+from copy import deepcopy
+from dataclasses import dataclass, field
+from datetime import datetime
+from math import ceil
+from typing import Optional, Callable, Any
+
+from goodai.helpers.json_helper import sanitize_and_parse_json, SimpleJSONEncoder, SimpleJSONDecoder
+from goodai.ltm.mem.auto import AutoTextMemory
+from goodai.ltm.mem.config import TextMemoryConfig
+from goodai.ltm.mem.default import DefaultTextMemory
+from litellm import token_counter
+import ltm.scratchpad as sp
+from model_interfaces.base_ltm_agent import Message
+
+from utils.llm import LLMContext, make_system_message, make_user_message, make_assistant_message, ensure_context_len, \
+    ask_llm
+from utils.text import truncate
+from utils.ui import colour_print
+
+
+def dump_context_s(context):
+    messages = []
+    for message in context:
+        messages.append(f"---\nRole: {message['role']}\nContent:\n{message['content']}")
+
+    return "\n".join(messages)
+
+@dataclass
+class InsertedContextAgent:
+    max_prompt_size: int
+    max_completion_tokens: Optional[int] = None
+    semantic_memory: DefaultTextMemory = field(default_factory=AutoTextMemory.create)
+    max_prompt_size: int = 16384
+    is_local: bool = True
+    defined_kws: list = field(default_factory=list)
+    scratchpad: dict[str, str] = field(default_factory=dict)
+    scratchpad_ts: dict[str, float] = field(default_factory=dict)  # Tracks changes in user_info (timestamps)
+    llm_index: int = 0
+    costs_usd: float = 0.0
+    model: str = "together_ai/meta-llama/Meta-Llama-3.1-70B-Instruct-Turbo"
+    temperature: float = 0.01
+    system_message = """You are a helpful AI assistant."""
+
+    def __post_init__(self):
+        self.semantic_memory = AutoTextMemory.create(config=TextMemoryConfig(chunk_capacity=50, chunk_overlap_fraction=0.0))
+        self.is_local = True
+        self.max_message_size = 1000
+        self.defined_kws = []
+        self.new_session()
+
+    def new_session(self) -> 'LTMAgentSession':
+        """
+        Creates a new LTMAgentSession object and sets it as the current session.
+        :return: The new session object.
+        """
+        session_id = str(uuid.uuid4())
+        self.session = LTMAgentSession(session_id=session_id, m_history=[], i_dict={})
+        if not self.semantic_memory.is_empty():
+            self.semantic_memory.add_separator()
+        return self.session
+
+    def reply(self, user_message: str, agent_response: Optional[str] = None, cost_callback: Callable=None) -> str:
```

Comment:
There are a couple of things you can leverage from the current baseline:
1. Set `self.now` at the very beginning of `reply`, to have a consistent reference timestamp.
2. Use `self._completion` to clean up the code a bit.
3. The function `td_format` to show relative timestamps.
(and probably other small things that I'm forgetting now)